src/bson.rs

```diff
@@ -3,3 +3,82 @@
 // Because BSON depends on the protocol version,
 // need to develop an Enum<Struct, Struct, ...> to
 // work based on the protocol version.
+use bson::spec::BinarySubtype;
+use bson::{bson, Binary, Bson, Document};
+use serde::ser::{Serialize, SerializeStruct, Serializer};
+use serde::Deserialize;
+use std::io::{Error, ErrorKind};
+
+#[derive(Deserialize, PartialEq, Debug)]
+struct AccVersion1 {
+    v: i32,
+    //other stuff
+    d: bson::Binary,
+}
+impl AccVersion1 {
+    fn new(data: Vec<u8>) -> AccVersion1 {
+        AccVersion1 {
+            v: 1,
+            d: bson::Binary {
+                subtype: bson::spec::BinarySubtype::Generic,
+                bytes: data,
+            },
+        }
+    }
+}
+impl Serialize for AccVersion1 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("AccVersion1", 2)?;
+        state.serialize_field("v", &self.v)?;
+        state.serialize_field("d", &self.d)?;
+        state.end()
+    }
+}
+#[derive(PartialEq, Debug)]
+enum PacketVersion {
+    V1(AccVersion1),
+}
+
+fn encapsulate(acc_enum: &PacketVersion) -> Result<Vec<u8>, std::io::Error> {
+    let acc_struct = match acc_enum {
+        PacketVersion::V1(structure) => structure,
+        _ => return Err(Error::from(ErrorKind::InvalidData)),
+    };
+    Ok(bson::to_vec(&acc_struct).unwrap())
+}
+
+fn reveal(bson_vec: Vec<u8>) -> Result<PacketVersion, std::io::Error> {
+    let doc: Document = bson::from_slice(&bson_vec[..]).unwrap();
```

Comment:
Can BSON not take `&bson_vec` as a slice? `Vec<u8>` should implement `AsRef<&[u8]>`